File: go/unique_binary_search_trees.go
func numTrees(n int) int {
    dp := []int{1, 1, 2}
    if n < 3 {
        return dp[n]
    }
    
    for i := 3; i < n+1; i++ {
        curr := 0
        for j := 1; j < i+1; j++ {
            curr += dp[j-1]*dp[i-j]
        }
        dp = append(dp, curr)
    }
    return dp[len(dp)-1]
}

File: python/unique-binary-search-trees.py
class Solution:
    def numTrees(self, n: int) -> int:
        if n < 3: return n
        dp = [1, 1, 2]
        
        for i in range(3, n+1):
            curr = 0
            for j in range(1, i+1):
                curr += dp[j-1]*dp[i-j]
            dp.append(curr)
        return dp[-1]